raster/src/main/scala/geotrellis/raster/GridBounds.scala

```diff
@@ -18,70 +18,20 @@ package geotrellis.raster
 
 import scala.collection.mutable
 import spire.syntax.cfor._
-
-/**
-  * The companion object for the [[GridBounds]] type.
-  */
-object GridBounds {
-  /**
-    * Given a [[CellGrid]], produce the corresponding [[GridBounds]].
-    *
-    * @param  r  The given CellGrid
-    */
-  def apply(r: CellGrid): GridBounds =
-    GridBounds(0, 0, r.cols-1, r.rows-1)
-
-  /**
-    * Given a sequence of keys, return a [[GridBounds]] of minimal
-    * size which covers them.
-    *
-    * @param  keys  The sequence of keys to cover
-    */
-  def envelope(keys: Iterable[Product2[Int, Int]]): GridBounds = {
-    var colMin = Integer.MAX_VALUE
-    var colMax = Integer.MIN_VALUE
-    var rowMin = Integer.MAX_VALUE
-    var rowMax = Integer.MIN_VALUE
-
-    for (key <- keys) {
-      val col = key._1
-      val row = key._2
-      if (col < colMin) colMin = col
-      if (col > colMax) colMax = col
-      if (row < rowMin) rowMin = row
-      if (row > rowMax) rowMax = row
-    }
-    GridBounds(colMin, rowMin, colMax, rowMax)
-  }
-
-  /**
-    * Creates a sequence of distinct [[GridBounds]] out of a set of
-    * potentially overlapping grid bounds.
-    *
-    * @param  gridBounds  A traversable collection of GridBounds
-    */
-  def distinct(gridBounds: Traversable[GridBounds]): Seq[GridBounds] =
-    gridBounds.foldLeft(Seq[GridBounds]()) { (acc, bounds) =>
-      acc ++ acc.foldLeft(Seq(bounds)) { (cuts, bounds) =>
-        cuts.flatMap(_ - bounds)
-      }
-    }
-}
+import spire.math._
+import spire.implicits._
 
 /**
   * Represents grid coordinates of a subsection of a RasterExtent.
   * These coordinates are inclusive.
   */
-case class GridBounds(colMin: Int, rowMin: Int, colMax: Int, rowMax: Int) {
-  def width = colMax - colMin + 1
-  def height = rowMax - rowMin + 1
-
-  def size: Long = width.toLong * height.toLong
-
-  @deprecated("Use `size` instead.", "2.0")
-  def sizeLong: Long = width.toLong * height.toLong
-
-  def isEmpty = size == 0
+case class GridBounds[@specialized(Short, Int, Long) N: Integral](
```

Comment:
It might be sensible just to specialize on `Int`/`Long` to avoid byte code duplication for a type that we won't likely see